components/de_amp/browser/test/de_amp_browsertest.cc

```diff
@@ -0,0 +1,274 @@
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_commands.h"
+#include "chrome/browser/ui/browser_list.h"
+#include "chrome/test/base/in_process_browser_test.h"
+#include "chrome/test/base/ui_test_utils.h"
+#include "components/prefs/testing_pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/test/browser_test.h"
+#include "content/public/test/content_mock_cert_verifier.h"
+#include "content/public/test/test_navigation_observer.h"
+#include "net/dns/mock_host_resolver.h"
+#include "third_party/re2/src/re2/re2.h"
+
+const char kTestHost[] = "a.test.com";
+const char kTestAmpPage[] = "/test.html";
+const char kTestSimpleNonAmpPage[] = "/simple.html";
+const char kTestCanonicalPage[] = "/simple_canonical.html";
+const char kTestBody[] =
+    R"(
+      <html amp>
+    <head>
+    <link rel='canonical'
+    href='https://%s:%s%s'>
+    </head>
+    </html>
+    )";
+class DeAmpBrowserTest : public InProcessBrowserTest {
+ public:
+  DeAmpBrowserTest() {
+    feature_list_.InitAndEnableFeature(de_amp::features::kBraveDeAMP);
+  }
+
+  void SetUpOnMainThread() override {
+    mock_cert_verifier_.mock_cert_verifier()->set_default_result(net::OK);
+    host_resolver()->AddRule("*", "127.0.0.1");
+    https_server_.reset(new net::EmbeddedTestServer(
+        net::test_server::EmbeddedTestServer::TYPE_HTTPS));
+
+    // Serve test files
+    base::FilePath path;
+    base::PathService::Get(base::DIR_SRC_TEST_DATA_ROOT, &path);
+    path = path.Append(FILE_PATH_LITERAL("brave"));
+    path = path.Append(FILE_PATH_LITERAL("components"));
+    path = path.Append(FILE_PATH_LITERAL("de_amp"));
+    path = path.Append(FILE_PATH_LITERAL("browser"));
+    path = path.Append(FILE_PATH_LITERAL("test"));
+    path = path.Append(FILE_PATH_LITERAL("data"));
+    https_server_->ServeFilesFromDirectory(path);
+
+    prefs_ = browser()->profile()->GetPrefs();
+
+    content::SetupCrossSiteRedirector(https_server_.get());
+    InProcessBrowserTest::SetUpOnMainThread();
+  }
+
+  void SetUpCommandLine(base::CommandLine* command_line) override {
+    InProcessBrowserTest::SetUpCommandLine(command_line);
+    mock_cert_verifier_.SetUpCommandLine(command_line);
+  }
+
+  void SetUpInProcessBrowserTestFixture() override {
+    InProcessBrowserTest::SetUpInProcessBrowserTestFixture();
+    mock_cert_verifier_.SetUpInProcessBrowserTestFixture();
+  }
+
+  void TearDownInProcessBrowserTestFixture() override {
+    InProcessBrowserTest::TearDownInProcessBrowserTestFixture();
+    mock_cert_verifier_.TearDownInProcessBrowserTestFixture();
+  }
+
+  content::WebContents* web_contents() {
+    return browser()->tab_strip_model()->GetActiveWebContents();
+  }
+
+  void TogglePref(const bool on) {
+    prefs_->SetBoolean(de_amp::kDeAmpPrefEnabled, on);
+    web_contents()->GetController().Reload(content::ReloadType::NORMAL, false);
+  }
+
+  void NavigateToURLAndWaitForRedirects(const GURL& original_url,
```

Comment:
Chromium has a similar thing, can we reuse it?
https://source.chromium.org/chromium/chromium/src/+/main:content/public/test/browser_test_utils.h;l=138-145;drc=4d2b5adb556128aab9313fc3851f192c254e09cb;bpv=1;bpt=1

Reply:
This doesn't work when I try it locally, it only ever sees the original URL. Perhaps because the navigation is cancelled?

Reply:
That method only waits for one navigation to finish before checking the last committed URL: https://source.chromium.org/chromium/chromium/src/+/main:content/public/test/browser_test_utils.cc;l=640;drc=5a758a97032f0b656c3c36a3497560762495501a;bpv=1;bpt=1